### PR TITLE
Use a database for mod storage

### LINKED ISSUE #10978: Add database backends for mod_storage
## Problem

Mod storage is currently stored using JSON files (one per mod). This is bad for performance if mod storage grows significantly. Mods like sban have accordingly started to require their own database backend.

## Solutions

Convert mod storage to use the available database backend for performance boost.

## Alternatives

None. We have a database, so we should use it for the simple key-value mod storage.

### PR DESCRIPTION
**Pull request by TurkeyMcMac**
Add compact, short information about your PR for easier understanding:

- Goal of the PR
  - Use a database instead of JSON files for mod storage.
- How does the PR work?
  - I added the abstract class `ModMetadataDatabase` to `src/database/database.h`. Backends are implemented for `sqlite3`, `files`, and `dummy`. `files` is backwards-compatible with the old format.
  - Client-side storage uses SQLite as its backend. This is a breaking change, but the client-side API is unstable.
- Does it resolve any reported issue?
  - It would fix #10978.
- Does this relate to a goal in [the roadmap](../doc/direction.md)?
  - This was a suggestion by the modder @appgurueu.
- If not a bug fix, why is this PR needed? What usecases does it solve?
  - #10978 contains some rationale.

## To do

This PR is ready for review.

## How to test

There are now unit tests for mod storage. You can also try out the database backend with mods that use storage. To use the SQLite3 backend in a preexisting world, first migrate to it with `--migrate-mod-storage sqlite3`.

### CODE REVIEW

File: src/database/database-dummy.cpp
Comment (appgurueu) [inline src/database/database-dummy.cpp:84]: Shouldn't all of these work in-memory (currently they are no-op)?

File: src/server.cpp
Comment (appgurueu) [inline src/server.cpp:750]: Can't you drop the entire loop then?

Comment (TurkeyMcMac) [reply src/server.cpp:750]: I haven't gotten rid of all the unneeded code yet.

File: src/database/database-dummy.cpp
Comment (TurkeyMcMac) [reply src/database/database-dummy.cpp:84]: Yes. Right now these no-ops work fine because of the way `ModMetadata` interacts with the database (it keeps its own data in-memory.)

File: src/database/database-sqlite3.cpp
Comment (appgurueu) [inline src/database/database-sqlite3.cpp:802]: Isn't this redundant due to the primary key constraint below?

Comment (TurkeyMcMac) [reply src/database/database-sqlite3.cpp:802]: Apparently not: https://sqlite.org/lang_createtable.html#the_primary_key

**Comment by TurkeyMcMac** [comment]
Is there a way to run the unit tests on Android?

**Comment by appgurueu** [comment]
@sfan5 I would consider this a (long overdue) performance improvement rather than a feature

File: src/database/database-sqlite3.cpp
Comment (rubenwardy) [inline src/database/database-sqlite3.cpp:800]: why "pairs"? Is this term used elsewhere? "entry" would make more sense, it's more commonly used by key-value dictionaries

File: src/database/database-sqlite3.cpp
Comment (rubenwardy) [inline src/database/database-sqlite3.cpp:801]: This should be a `VARCHAR(30)` or so, does Minetest have a set limit on mod names?

File: src/database/database-sqlite3.cpp
Comment (rubenwardy) [inline src/database/database-sqlite3.cpp:802]: Why is this a blob and not a VARCHAR or TEXT?

File: src/database/database-sqlite3.cpp
Comment (TurkeyMcMac) [reply src/database/database-sqlite3.cpp:800]: Good point. I will change the name of the table to "entries".

File: src/database/database-sqlite3.cpp
Comment (appgurueu) [reply src/database/database-sqlite3.cpp:802]: Because Lua strings are byte strings. A BLOB best represents that.

File: src/database/database-sqlite3.cpp
Comment (TurkeyMcMac) [reply src/database/database-sqlite3.cpp:801]: Apparently SQLite treats `VARCHAR(N)` exactly the same as `TEXT` (it doesn't enforce the limit.) I could just change the type `VARCHAR(64)` or something.

Comment (appgurueu) [reply src/database/database-sqlite3.cpp:801]: > This should be a `VARCHAR(30)` or so, does Minetest have a set limit on mod names?

It should not be a `VARCHAR(30)` for (1) future flexibility (who will remember to change this if the mod name length limit is bumped? a constant would have be introduced) and (2) SQLite ignores the type arguments anyways:
> Note that numeric arguments in parentheses that following the type name (ex: "VARCHAR(255)") are ignored by SQLite - SQLite does not impose any length restrictions (other than the large global SQLITE_MAX_LENGTH limit) on the length of strings, BLOBs or numeric values.

File: src/database/database-sqlite3.cpp
Comment (appgurueu) [reply src/database/database-sqlite3.cpp:800]: I'm leaning towards `entry` too, but considering that the Lua table iterator is called `pairs` and not `entries`, I consider `pairs` fine as more Lua-esque terminology.

File: src/database/database-sqlite3.cpp
Comment (TurkeyMcMac) [reply src/database/database-sqlite3.cpp:802]: If the key and value are guaranteed to be valid UTF-8, I can change the column types to TEXT.

If the key and value are intended to be limited in length, I can change the column types to VARCHAR.

File: src/database/database-sqlite3.cpp
Comment (TurkeyMcMac) [reply src/database/database-sqlite3.cpp:800]: If I change the name of the table, should I call it "entry" or "entries"? The plural form seems to be consistent with most of the other SQLite table names used.

**Comment by TurkeyMcMac** [comment]
Commit 37da46c changes a table name. You will have to modify or remove both the server-side and client-side SQLite mod metadata databases for things to work.

**Comment by Emojigit** [comment]
Can I migrate from file to sqlite3?

**Comment by TurkeyMcMac** [comment]
@Emojigit Yes. You should just be able to use the migration option `--migrate-mod-metadata sqlite3` like you would other migration options. Beware that the format still may change somewhat before this PR is merged.

File: src/main.cpp
Comment (SmallJoker) [inline src/main.cpp:303]: I think it would be more comprehensive if this option were called `mod storage` rather than `mod metadata` even if the current naming is technically correct.

File: src/database/database-sqlite3.cpp
Comment (SmallJoker) [inline src/database/database-sqlite3.cpp:880]: Wouldn't it be more efficient to use the mod name as table name, rather than a separate column?

**Comment by SmallJoker** [comment]
Migrated the database back and forth. Here's my test code to initialize the mod storage with some dummy values which then can be checked for validity after migration.
```Lua
local m = minetest.get_mod_storage()

local DO_WRITE = false
local DO_CHECK = (not DO_WRITE) and true

if DO_WRITE then
	-- Before migration step
	m:from_table() -- Clear
	m:from_table({
		fields = {
			string1 = "hello",
			int1 = 420,
			float = 3.141
		}
	})
	m:set_int("int2", m:get_int("int1") + 100)
	m:set_string("string2", "world")
	m:set_string("string1", "") -- Clear
	m:set_float("float", 0.707)
end

if DO_CHECK then
	-- After migration step
	assert(m:to_table().fields.string2 == "world")
	assert(m:get_int("int2") == 520)
	assert(math.abs(m:get_float("float") - 0.707) < 0.0001)
end

if not DO_CHECK and not DO_WRITE then
core.register_chatcommand("store", {
	func = function(name, param)
		local args = param:split(" ", false, 2)
		if args[1] == "get" and #args == 2 then
			return m:contains(args[2]), m:contains(args[2]) and m:get(args[2]) or "NOT FOUND"
		end
		if args[1] == "set" and #args == 3 then
			m:set_string(args[2], args[3])
			return true, "OK"
		end
		if args[1] == "rm" and #args == 2 then
			m:set_string(args[2], "")
			return true, "OK"
		end
		return false, "Format: /store [get|set|rm] key [value]"
	end
})
end
```

Works.

File: src/database/database-sqlite3.cpp
Comment (TurkeyMcMac) [reply src/database/database-sqlite3.cpp:880]: It may be. I haven't tested. The currently implementation is simpler, especially within the existing framework of `Database_SQLite3`. IMO the important thing is that the cost of changing a metadata entry is now dependent on the size of the entry rather than the combined size of all the mod's entries.

File: src/main.cpp
Comment (TurkeyMcMac) [reply src/main.cpp:303]: Should I also change the `mod_metadata_backend` config option?

Comment (SmallJoker) [reply src/main.cpp:303]: Yes, I think so for the same reason as above that "mod storage" is more known than "mod metadata".

**Comment by TurkeyMcMac** [comment]
A recent commit changes the setting `mod_metadata_backend` to `mod_storage_backend`. If you have any `world.mt` files with the old setting present, update them accordingly.

File: src/script/lua_api/l_storage.cpp
Comment (SmallJoker) [inline src/script/lua_api/l_storage.cpp:37]: `store` must be deleted if this returns `false`.

**Review by SmallJoker — APPROVED** [review]
Thank you for the changes.

**Review by celeron55 — CHANGES_REQUESTED** [review]
I'll approve once the weird error message says something even remotely correct. Otherwise code looks fine and tests ok. This definitely is a needed feature.

File: src/database/database.h
Comment (celeron55) [reply src/database/database.h:28]: For some reason Github doesn't delete the above comment when I tell it to delete it.

File: src/database/database-sqlite3.cpp
Comment (TurkeyMcMac) [reply src/database/database-sqlite3.cpp:847]: Sorry, that must be a copy-paste issue.

**Comment by TurkeyMcMac** [comment]
Before this is merged, the unit tests should probably be run on Android. The SQLite3 backend uses `REPLACE INTO`, which may not work on Android, judging by the workaround used in the SQLite3 map backend: https://github.com/minetest/minetest/blob/4fd6a36d6576a89bd54d48fed0322542141f3744/src/database/database-sqlite3.cpp#L231-L235

**Review by celeron55 — COMMENTED** [review]
I'll leave my review unapproved to hilight the fact that this can't be merged before it has been tested on Android.

**Comment by TurkeyMcMac** [comment]
I tested this on my Android phone with a mod that uses mod storage. The `REPLACE INTO` statement works; stored entries are successfully replaced.

**Comment by celeron55** [comment]
What about Android versions? There must be some reason for the database-sqlite3.cpp #ifdef.

**Comment by rubenwardy** [comment]
The ifdef was added in https://github.com/minetest/minetest/pull/1457
No reason given. Shoddy programming

It's probably worth checking the IRC logs around that time. I'm going to go for lunch though

**Comment by celeron55** [comment]
https://irc.minetest.net/minetest-dev/2014-07-08

**Comment by celeron55** [comment]
I might go as far as to say if it experimentally works now, it's good without the #ifdef. It's been 8 years and nobody uses the same Android versions anymore.

**Comment by appgurueu** [comment]
> I might go as far as to say if it experimentally works now, it's good without the #ifdef. It's been 8 years and nobody uses the same Android versions anymore. But it probably should be tested with something like Android 8.

I'd say the #ifdef should be reverted then.

**Review by celeron55 — APPROVED** [review]
As it works on Android 8.1, it's fine by me.

**Comment by celeron55** [comment]
It seems to me that if it fails to work for an old Android version, it will raise an exception and exit the game, showing an error message dialog. Now, I'm pretty sure we officially support older android than API 27, so doing a release without testing an older version is a bit reckless.

If this is merged, we need to add a blocking issue for the next release requesting testing of sqlite3 mod storage on older Android versions.

**Comment by appgurueu** [comment]
> EDIT: The other option is of course to just add the #ifdef and think about this some other time.

Not an option IMO. `REPLACE` cannot be replaced this way. In the PR ruben linked, it is noted that this is a dirty hack violating the "atomic" part of the ACID transaction principle.

**Comment by TurkeyMcMac** [comment]
By the way, it seems that `REPLACE INTO` is also used in the rollback code: https://github.com/minetest/minetest/blob/4fd6a36d6576a89bd54d48fed0322542141f3744/src/rollback.cpp#L290-L291

**Comment by celeron55** [comment]
> By the way, it seems that `REPLACE INTO` is also used in the rollback code:
> https://github.com/minetest/minetest/blob/4fd6a36d6576a89bd54d48fed0322542141f3744/src/rollback.cpp#L290-L291

This code was written in 2014 just like #1457, almost at the same time. But I wouldn't expect it to having been tested on Android back then.

**Comment by appgurueu** [comment]
> This code was written in 2014 just like #1457, almost at the same time. But I wouldn't expect it to have been tested on Android back then.

I wouldn't expect Android users to have rollback enabled at all.

**Comment by celeron55** [comment]
>If this is merged, we need to add a blocking issue for the next release requesting testing of sqlite3 mod storage on older Android versions.

I would like to proceed with this to avoid rebasing perfectly good code. Is it ok or do people feel 5.5 already has too many blockers?

**Comment by rubenwardy** [comment]
Sounds fine by me

**Pull request merged**
